File: concepts/03-reliability/python/health_checks.py
"""
Health Checks Implementation
Demonstrates liveness, readiness, and startup probes
"""
import asyncio
import random
from typing import Dict, Any
from datetime import datetime


class HealthChecker:
    """Health checker for service health monitoring"""
    
    def __init__(self):
        self.is_alive = True
        self.is_ready = False
        self.is_started = False
        self.dependencies = {
            'database': False,
            'cache': False,
            'external_service': False
        }
        self.start_time = datetime.now()
        self.initialization_time = 2.0  # 2 seconds
    
    async def initialize(self):
        """Initialize service (simulate startup)"""
        print('Initializing service...')
        
        # Simulate initialization
        await asyncio.sleep(self.initialization_time)
        
        # Check dependencies
        await self.check_dependencies()
        
        self.is_started = True
        self.is_ready = True
        print('Service initialized and ready')
    
    async def check_dependencies(self) -> Dict[str, bool]:
        """Check dependencies"""
        # Simulate dependency checks
        self.dependencies['database'] = random.random() > 0.2  # 80% chance
        self.dependencies['cache'] = random.random() > 0.1  # 90% chance
        self.dependencies['external_service'] = random.random() > 0.15  # 85% chance
        
        return self.dependencies
    
    def liveness_probe(self) -> Dict[str, Any]:
        """Liveness Probe - checks if service is alive"""
        if not self.is_alive:
            return {
                'status': 'unhealthy',
                'message': 'Service is not alive',
                'timestamp': datetime.now().isoformat()
            }
        
        # Check if process is responsive
        uptime = (datetime.now() - self.start_time).total_seconds()
        
        return {
            'status': 'healthy',
            'message': 'Service is alive',
            'uptime': f'{int(uptime)}s',
            'timestamp': datetime.now().isoformat()
        }
    
    def readiness_probe(self) -> Dict[str, Any]:
        """Readiness Probe - checks if service is ready to serve traffic"""
        if not self.is_ready:
            return {
                'status': 'not ready',
                'message': 'Service is not ready to serve traffic',
                'dependencies': self.dependencies,
                'timestamp': datetime.now().isoformat()
            }
        
        # Check if all critical dependencies are available
        critical_deps = ['database', 'cache']
        unavailable_deps = [dep for dep in critical_deps if not self.dependencies[dep]]
        
        if unavailable_deps:
            return {
                'status': 'not ready',
                'message': 'Critical dependencies unavailable',
                'unavailable_dependencies': unavailable_deps,
                'dependencies': self.dependencies,
                'timestamp': datetime.now().isoformat()
            }
        
        return {
            'status': 'ready',
            'message': 'Service is ready to serve traffic',
            'dependencies': self.dependencies,
            'timestamp': datetime.now().isoformat()
        }
    
    def startup_probe(self) -> Dict[str, Any]:
        """Startup Probe - checks if service has finished starting up"""
        if self.is_started:
            return {
                'status': 'started',
                'message': 'Service has started',
                'initialization_time': f'{self.initialization_time * 1000}ms',
                'timestamp': datetime.now().isoformat()
            }
        
        elapsed = (datetime.now() - self.start_time).total_seconds() * 1000
        remaining = max(0, self.initialization_time * 1000 - elapsed)
        
        return {
            'status': 'starting',
            'message': 'Service is still starting',
            'elapsed': f'{int(elapsed)}ms',
            'estimated_remaining': f'{int(remaining)}ms',
            'timestamp': datetime.now().isoformat()
        }
    
    def health_check(self) -> Dict[str, Any]:
        """Combined health check"""
        return {
            'liveness': self.liveness_probe(),
            'readiness': self.readiness_probe(),
            'startup': self.startup_probe(),
            'overall': self.get_overall_health()
        }
    
    def get_overall_health(self) -> str:
        """Get overall health status"""
        liveness = self.liveness_probe()
        readiness = self.readiness_probe()
        startup = self.startup_probe()
        
        if liveness['status'] != 'healthy':
            return 'unhealthy'
        
        if startup['status'] == 'starting':
            return 'starting'
        
        if readiness['status'] != 'ready':
            return 'not ready'
        
        return 'healthy'
    
    def simulate_failure(self):
        """Simulate service failure"""
        self.is_alive = False
        print('Service failure simulated')
    
    def simulate_dependency_failure(self, dependency: str):
        """Simulate dependency failure"""
        if dependency in self.dependencies:
            self.dependencies[dependency] = False
            self.is_ready = False
            print(f'Dependency failure simulated: {dependency}')
    
    def recover(self):
        """Recover from failure"""
        self.is_alive = True
        self.is_ready = True
        for key in self.dependencies:
            self.dependencies[key] = True
        print('Service recovered')


class HealthCheckServer:
    """Health check server with endpoints"""
    
    def __init__(self, health_checker: HealthChecker, port: int = 3000):
        self.health_checker = health_checker
        self.port = port
    
    def start(self):
        """Start health check endpoints"""
        print(f'Health check server started on port {self.port}')
        print('Endpoints:')
        print('  GET /health - Combined health check')
        print('  GET /live - Liveness probe')
        print('  GET /ready - Readiness probe')
        print('  GET /startup - Startup probe')
    
    def handle_request(self, path: str) -> Dict[str, Any]:
        """Handle health check request"""
        if path == '/health':
            return self.health_checker.health_check()
        elif path == '/live':
            return self.health_checker.liveness_probe()
        elif path == '/ready':
            return self.health_checker.readiness_probe()
        elif path == '/startup':
            return self.health_checker.startup_probe()
        else:
            return {'error': 'Not found'}


async def demonstrate_health_checks():
    """Demonstrate health checks"""
    print('=== Health Checks Demonstration ===\n')
    
    health_checker = HealthChecker()
    server = HealthCheckServer(health_checker)
    
    # Initial state (starting)
    print('1. Initial state (starting):')
    import json
    print(json.dumps(health_checker.startup_probe(), indent=2))
    
    # Initialize service
    await health_checker.initialize()
    
    # After initialization
    print('\n2. After initialization:')
    print(json.dumps(health_checker.health_check(), indent=2))
    
    # Simulate dependency failure
    print('\n3. Simulating dependency failure:')
    health_checker.simulate_dependency_failure('database')
    print(json.dumps(health_checker.readiness_probe(), indent=2))
    
    # Recover
    print('\n4. Recovering:')
    health_checker.recover()
    print(json.dumps(health_checker.health_check(), indent=2))
    
    # Simulate service failure
    print('\n5. Simulating service failure:')
    health_checker.simulate_failure()
    print(json.dumps(health_checker.liveness_probe(), indent=2))


if __name__ == '__main__':
    asyncio.run(demonstrate_health_checks())

File: concepts/03-reliability/node/health-checks.js
/**
 * Health Checks Implementation
 * Demonstrates liveness, readiness, and startup probes
 */

class HealthChecker {
  constructor() {
    this.isAlive = true;
    this.isReady = false;
    this.isStarted = false;
    this.dependencies = {
      database: false,
      cache: false,
      externalService: false
    };
    this.startTime = Date.now();
    this.initializationTime = 2000; // 2 seconds
  }

  /**
   * Initialize service (simulate startup)
   */
  async initialize() {
    console.log('Initializing service...');
    
    // Simulate initialization
    await new Promise(resolve => setTimeout(resolve, this.initializationTime));
    
    // Check dependencies
    await this.checkDependencies();
    
    this.isStarted = true;
    this.isReady = true;
    console.log('Service initialized and ready');
  }

  /**
   * Check dependencies
   */
  async checkDependencies() {
    // Simulate dependency checks
    this.dependencies.database = Math.random() > 0.2; // 80% chance of success
    this.dependencies.cache = Math.random() > 0.1; // 90% chance of success
    this.dependencies.externalService = Math.random() > 0.15; // 85% chance of success
    
    return this.dependencies;
  }

  /**
   * Liveness Probe
   * Checks if the service is alive (not deadlocked, crashed, etc.)
   */
  livenessProbe() {
    if (!this.isAlive) {
      return {
        status: 'unhealthy',
        message: 'Service is not alive',
        timestamp: new Date().toISOString()
      };
    }

    // Check if process is responsive
    const uptime = Date.now() - this.startTime;
    
    return {
      status: 'healthy',
      message: 'Service is alive',
      uptime: `${Math.floor(uptime / 1000)}s`,
      timestamp: new Date().toISOString()
    };
  }

  /**
   * Readiness Probe
   * Checks if the service is ready to serve traffic
   */
  readinessProbe() {
    if (!this.isReady) {
      return {
        status: 'not ready',
        message: 'Service is not ready to serve traffic',
        dependencies: this.dependencies,
        timestamp: new Date().toISOString()
      };
    }

    // Check if all critical dependencies are available
    const criticalDeps = ['database', 'cache'];
    const unavailableDeps = criticalDeps.filter(dep => !this.dependencies[dep]);

    if (unavailableDeps.length > 0) {
      return {
        status: 'not ready',
        message: 'Critical dependencies unavailable',
        unavailableDependencies: unavailableDeps,
        dependencies: this.dependencies,
        timestamp: new Date().toISOString()
      };
    }

    return {
      status: 'ready',
      message: 'Service is ready to serve traffic',
      dependencies: this.dependencies,
      timestamp: new Date().toISOString()
    };
  }

  /**
   * Startup Probe
   * Checks if the service has finished starting up
   */
  startupProbe() {
    if (this.isStarted) {
      return {
        status: 'started',
        message: 'Service has started',
        initializationTime: `${this.initializationTime}ms`,
        timestamp: new Date().toISOString()
      };
    }

    const elapsed = Date.now() - this.startTime;
    const remaining = Math.max(0, this.initializationTime - elapsed);

    return {
      status: 'starting',
      message: 'Service is still starting',
      elapsed: `${elapsed}ms`,
      estimatedRemaining: `${remaining}ms`,
      timestamp: new Date().toISOString()
    };
  }

  /**
   * Combined health check
   */
  healthCheck() {
    return {
      liveness: this.livenessProbe(),
      readiness: this.readinessProbe(),
      startup: this.startupProbe(),
      overall: this.getOverallHealth()
    };
  }

  /**
   * Get overall health status
   */
  getOverallHealth() {
    const liveness = this.livenessProbe();
    const readiness = this.readinessProbe();
    const startup = this.startupProbe();

    if (liveness.status !== 'healthy') {
      return 'unhealthy';
    }

    if (startup.status === 'starting') {
      return 'starting';
    }

    if (readiness.status !== 'ready') {
      return 'not ready';
    }

    return 'healthy';
  }

  /**
   * Simulate service failure
   */
  simulateFailure() {
    this.isAlive = false;
    console.log('Service failure simulated');
  }

  /**
   * Simulate dependency failure
   */
  simulateDependencyFailure(dependency) {
    if (this.dependencies.hasOwnProperty(dependency)) {
      this.dependencies[dependency] = false;
      this.isReady = false;
      console.log(`Dependency failure simulated: ${dependency}`);
    }
  }

  /**
   * Recover from failure
   */
  recover() {
    this.isAlive = true;
    this.isReady = true;
    Object.keys(this.dependencies).forEach(key => {
      this.dependencies[key] = true;
    });
    console.log('Service recovered');
  }
}

/**
 * Health Check Server
 */
class HealthCheckServer {
  constructor(healthChecker, port = 3000) {
    this.healthChecker = healthChecker;
    this.port = port;
  }

  /**
   * Start health check endpoints
   */
  start() {
    // Simulate HTTP server endpoints
    console.log(`Health check server started on port ${this.port}`);
    console.log('Endpoints:');
    console.log('  GET /health - Combined health check');
    console.log('  GET /live - Liveness probe');
    console.log('  GET /ready - Readiness probe');
    console.log('  GET /startup - Startup probe');
  }

  /**
   * Handle health check request
   */
  handleRequest(path) {
    switch (path) {
      case '/health':
        return this.healthChecker.healthCheck();
      case '/live':
        return this.healthChecker.livenessProbe();
      case '/ready':
        return this.healthChecker.readinessProbe();
      case '/startup':
        return this.healthChecker.startupProbe();
      default:
        return { error: 'Not found' };
    }
  }
}

// Example usage
async function demonstrateHealthChecks() {
  console.log('=== Health Checks Demonstration ===\n');

  const healthChecker = new HealthChecker();
  const server = new HealthCheckServer(healthChecker);

  // Initial state (starting)
  console.log('1. Initial state (starting):');
  console.log(JSON.stringify(healthChecker.startupProbe(), null, 2));

  // Initialize service
  await healthChecker.initialize();

  // After initialization
  console.log('\n2. After initialization:');
  console.log(JSON.stringify(healthChecker.healthCheck(), null, 2));

  // Simulate dependency failure
  console.log('\n3. Simulating dependency failure:');
  healthChecker.simulateDependencyFailure('database');
  console.log(JSON.stringify(healthChecker.readinessProbe(), null, 2));

  // Recover
  console.log('\n4. Recovering:');
  healthChecker.recover();
  console.log(JSON.stringify(healthChecker.healthCheck(), null, 2));

  // Simulate service failure
  console.log('\n5. Simulating service failure:');
  healthChecker.simulateFailure();
  console.log(JSON.stringify(healthChecker.livenessProbe(), null, 2));
}

if (require.main === module) {
  demonstrateHealthChecks();
}

module.exports = { HealthChecker, HealthCheckServer };

